### PR TITLE
Work in progress: compute contig alignments and call variants

### PR DESCRIPTION
**Pull request by standage**
Using ksw2 library. Against my better judgement, first thought is to implement as a C extension with CPython.

### CODE REVIEW

**Comment by ctb** [comment]
Random question: is this something that could usefully be separated from kevlar at some point in the future and made part of ksw2 (or at least a separate project)?

**Comment by standage** [comment]
@ctb Perhaps. Exposing ksw2 via Python bindings would require quite a bit more code, to support the different alignment algorithms, parameters, etc. In contrast, kevlar will be hard coding many of these (for now at least), so I can get away with minimal inter-language glue code. I hope.

**Comment by ctb** [comment]
ok!

**Comment by standage** [comment]
Dealing with the CPython boilerplate code and Python 2 / 3 compatibility was something I was not looking forward to. Cython made this much easier.                                                                          
                                                                                 
Also, there were problems with compiling my alignment code (which used C++) and the ksw2 code (which is pure C) together, so I dropped the convenience of std::string so that everything could compile nicely together in C.

**Comment by standage** [comment]
Good news:

- The new C extension compiles and the entire kevlar package installs without error.
- I can `import kevlar.align` with no problem once kevlar is installed. Oddly enough, I can also import `import align` without importing kevlar.

Bad news:

- Running `pip install .` takes about 3-5 minutes, during almost all of which the terminal produces no output. I've never seen behavior like this before.
- Neither the `kevlar.align` module (nor the unexpected `align` module) have any available functions (as inspected by `help(kevlar.align)` and `dir(kevlar.align)`).

Any ideas what's going on here? The extension is super simple, so I'm guessing I've made a dumb, simple mistake.

cc @camillescott @luizirber @betatim @ctb

**Comment by ctb** [comment]
if you pip install Cython first, the compile will go faster. I've never tracked it down but the long delay with no output seems to be from installing Cython impermanently.

**Comment by standage** [comment]
I had already installed cython before trying to build or install kevlar. Even in a fresh virtualenv (with cython pre-installed) the behavior was the same. :-/

But that's only a minor annoyance. The big question is why I can't invoke the `align` function.

**Comment by standage** [comment]
Ok, seems to be working now. There seemed to be a combination of issues with not fully exposing the function to Python land, as well as not importing properly, as well as transient caching issues on my machine.

Now it's building directly from the pre-cythonized C source code, so Cython shouldn't have to be a prerequisite.

Now I just need to figure out how to make sure the dev environment is always up-to-date.

**Comment by standage** [comment]
Ugh. [This](https://travis-ci.org/dib-lab/kevlar/jobs/255522020#L819) looks suspiciously like a text encoding problem. Works fine for python 2.7.

File: .travis.yml [hunk 1]
```diff
@@ -17,6 +17,7 @@ install:
     - pip install wheel
     - pip install pysam>=0.11.2
     - pip install -r requirements.txt
+    - pip install .
```
Comment (standage) [inline .travis.yml]: Even with adding the `env` make target as a prerequisite to `make testall`, the extension module wouldn't import properly until I added this line.

File: Makefile [hunk 2]
```diff
@@ -1,23 +1,41 @@
 SHELL=bash
 
+## #≠≠≠≠≠ build targets ≠≠≠≠≠#
+
+## help:     print this help message and exit
+help: Makefile
+	@sed -n 's/^## //p' Makefile
```
Comment (standage) [inline Makefile]: Make the build file self-documenting.

File: kevlar/tests/test_novel.py [hunk 3]
```diff
@@ -58,6 +58,7 @@ def test_assumptions(kmer):
     assert ct.get_kmer_hashes(kmer)[0] == ct.get_kmer_hashes(kmer_rc)[0]
 
 
+@pytest.mark.long
```
Comment (standage) [inline kevlar/tests/test_novel.py]: Uses deprecated code anyway, so no sense in taking the time to run it locally with each `make test` invocation.

File: setup.py [hunk 4]
```diff
@@ -7,10 +7,20 @@
...
 
-from setuptools import setup
+from setuptools import setup, Extension
 import versioneer
 
 
+align = Extension(
+    'kevlar.alignment',
+    sources=[
+        'kevlar/alignment.c', 'third-party/ksw2/ksw2_extz.c', 'src/align.c'
```
Comment (standage) [inline setup.py]: I initially had `kevlar/alignment.pyx` as the first source, but replaced it with `kevlar/alignment.c` after Cython did its thang. Any updates to the Cython code will require rebuilding the `kevlar/alignment.c` file.

File: setup.py [hunk 5]
```diff
@@ -25,7 +35,8 @@
           'kevlar': ['kevlar/tests/data/*', 'kevlar/tests/data/*/*']
       },
       include_package_data=True,
-      setup_requires=['pysam', 'networkx', 'pandas'],
+      ext_modules=[align],
+      setup_requires=['pysam', 'networkx', 'pandas', 'cython'],
```
Comment (standage) [inline setup.py]: TODO: drop cython as a setup dependency, add it to `make devenv` target.

Comment (standage) [inline setup.py]: See #92.

File: src/align.c [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    return int_seq;
+}
+
+
+// See align.h for documentation.
+void align(const char *target, const char *query, int match, int mismatch,
+           int gapopen, int gapextend, char *cigar)
+{
```
Comment (standage) [inline src/align.c]: Influenced heavily by the kws2 README.

**Pull request merged**